### PR TITLE
Refactor BGC data model

### PR DESCRIPTION
**Pull request by CunliangGeng**
Brief summary:
- remove MibigBGC class in order to use only BGC class for all types of BGC (mibig, antismash...)
- update BGC class
  - update and remove class parameters
  - remove or deprecate unused attributes
  - remove or deprecate unused methods
-  update BGC init process in BGC loaders
-  add docstrings and unit tests to updated classes

### CODE REVIEW

File: src/nplinker/genomics/gcf.py [hunk 1]
```diff
@@ -1,88 +1,40 @@
 from nplinker.strain_collection import StrainCollection
-from .mibig.mibig_bgc import MibigBGC
+from .bgc import BGC
```
Comment (hechth) [inline src/nplinker/genomics/gcf.py]: BGC and GCF import each other - is this addressed / solved somehow?

File: src/nplinker/genomics/bgc.py [hunk 2]
```diff
@@ -1,108 +1,146 @@
...
+        BGCs, e.g. antiSMASH has some self-defined features (like "region") in
+        its output GenBank files.
+
+        The annotations of BGC can be stored in JSON format, which is defined
+        and used by MIBiG.
+
+        Args:
+            bgc_id(str): BGC identifier, e.g. MIBiG accession, GenBank accession.
+            product_prediction(list[str]): list of BGC's (predicted) natural
+                products or product classes.
```
Comment (hechth) [inline src/nplinker/genomics/bgc.py:36]: Why the change from int to string?

File: src/nplinker/nplinker.py [hunk 3]
```diff
@@ -502,16 +502,16 @@ def get_common_strains(self, objects_a, objects_b, filter_no_shared=True):
...
-    def lookup_bgc(self, name):
-        """If BGC ``name`` exists, return it. Otherwise return None"""
-        if name not in self._bgc_lookup:
+    def lookup_bgc(self, bgc_id):
+        """If BGC ``bgc_id`` exists, return it. Otherwise return None"""
+        if bgc_id not in self._bgc_lookup:
             return None
 
-        return self._bgcs[self._bgc_lookup[name]]
+        return self._bgcs[self._bgc_lookup[bgc_id]]
```
Comment (hechth) [inline src/nplinker/nplinker.py]: Could be one line with ternary operator using method `has_bgc`

File: src/nplinker/scoring/rosetta/rosetta.py [hunk 4]
```diff
@@ -536,12 +536,12 @@ def export_to_csv(self, filename):
 
             csvwriter.writerow([
                 'nplinker spectrum ID', 'spectrum ID', 'GNPS ID',
-                'spectral score', 'nplinker BGC ID', 'BGC ID', 'MiBIG BGC ID',
+                'spectral score', 'BGC ID', 'MiBIG BGC ID',
```
Comment (hechth) [inline src/nplinker/scoring/rosetta/rosetta.py:539]: Should the column names include spaces?

**Review by hechth — COMMENTED** [review]
Looks good to me - so basically the `BGC` id and name were unified int a single attribute if I understand those changes correctly?

File: src/nplinker/genomics/gcf.py (hunk 1 above)
Comment (CunliangGeng) [reply src/nplinker/genomics/gcf.py]: yep, only import when doing type checking, see https://mypy.readthedocs.io/en/stable/runtime_troubles.html#import-cycles

File: src/nplinker/genomics/bgc.py (hunk 2 above)
Comment (CunliangGeng) [reply src/nplinker/genomics/bgc.py:36]: > so basically the BGC id and name were unified int a single attribute if I understand those changes correctly?

You're right, the former int `id` is useless and bgc `name` is more useful. To be consistent with other data models (e.g. GCF), here use string `bgc_id`, which is same as former `name`.

File: src/nplinker/nplinker.py (hunk 3 above)
Comment (CunliangGeng) [reply src/nplinker/nplinker.py]: Good idea! Updated for 3 lookup methods.

File: src/nplinker/scoring/rosetta/rosetta.py (hunk 4 above)
Comment (CunliangGeng) [reply src/nplinker/scoring/rosetta/rosetta.py:539]: it should be fine since it's csv file using comma as separator. I prefer not changing it at the moment.

**Comment by CunliangGeng** [comment]
> Looks good to me - so basically the `BGC` id and name were unified int a single attribute if I understand those changes correctly?

Thanks for the comments. You're right. The `id` in several classes (e.g. BGC, GCF) actually refers to index (index of one BGC object within a group of BGC objects), but not an independent and unique identifier of that object. It's quite confusing and also not really used. So I removed the former `id` and use `bgc_id` or `gcf_id` to refer to independent and unique identifier (which could be the former `name`).

Ready for 2nd round of review ;-)

**Comment by CunliangGeng** [comment]
The local test `test_nplinker.py` fails because of the change of parameter strain in BGC class. I'll solve it in a new PR, this issue is not only related to BGC but also the logics of loading and collecting strains.

**Review by hechth — APPROVED** [review]
